### ISSUE TITLE
Cannot boot from cdrom ISO

### ISSUE THREAD

**Issue by hassenius** [terraform-providers/terraform-provider-vsphere#644]
### Terraform Version

Terraform version: 0.11.7  41e50bd32a8825a84535e353c3674af8ce799161
+ provider.local v1.1.0
+ provider.null v1.0.0
+ provider.random v2.0.0
+ provider.template v1.0.0
+ provider.tls v1.2.0
+ provider.vsphere v1.8.1

### vSphere Provider Version
+ provider.vsphere v1.8.1

### Affected Resource(s)

- `vsphere_virtual_machine`

### Terraform Configuration Files
Relevant section
```hcl
resource "vsphere_virtual_machine" "icpmaster" {
  depends_on = ["vsphere_file.cloudinit_upload"]
  folder     = "${local.folder_path}"

  #####
  # VM Specifications
  ####
  count            = "${var.master["nodes"]}"
  resource_pool_id = "${data.vsphere_resource_pool.pool.id}"

  name      = "${format("${lower(var.instance_name)}-master%02d", count.index + 1) }"
  num_cpus  = "${var.master["vcpu"]}"
  memory    = "${var.master["memory"]}"

  scsi_controller_count = 1
  scsi_type = "lsilogic-sas"

  ####
  # Disk specifications
  ####
  datastore_id  = "${data.vsphere_datastore.datastore.id}"
  guest_id      = "${data.vsphere_virtual_machine.template.guest_id}"
  scsi_type     = "${data.vsphere_virtual_machine.template.scsi_type}"

  disk {
    label            = "${format("${lower(var.instance_name)}-master%02d.vmdk", count.index + 1) }"
    size             = "${var.master["disk_size"]        != "" ? var.master["disk_size"]        : data.vsphere_virtual_machine.template.disks.0.size}"
    eagerly_scrub    = "${var.master["eagerly_scrub"]    != "" ? var.master["eagerly_scrub"]    : data.vsphere_virtual_machine.template.disks.0.eagerly_scrub}"
    thin_provisioned = "${var.master["thin_provisioned"] != "" ? var.master["thin_provisioned"] : data.vsphere_virtual_machine.template.disks.0.thin_provisioned}"
    keep_on_remove   = "${var.master["keep_disk_on_remove"]}"
    unit_number      = 0
  }

  cdrom {
     datastore_id = "${data.vsphere_datastore.datastore.id}"
     path         = "cloud-init-iso/tests/${random_id.clusterid.dec}.iso"
  }
}

resource "vsphere_file" "cloudinit_upload" {
  depends_on       = ["null_resource.makeiso"]
  datacenter       = "${var.vsphere_datacenter}"
  datastore        = "${data.vsphere_datastore.datastore.id}"
  source_file      = "${path.module}/cloud-init/${random_id.clusterid.dec}.iso"
  destination_file = "cloud-init-iso/tests/${random_id.clusterid.dec}.iso"
  create_directories = true
}
```

### Debug Output
Relevant section:
```
2018-10-23T11:11:07.071+0100 [DEBUG] plugin.terraform-provider-vsphere_v1.8.1_x4: 2018/10/23 11:11:07 [DEBUG] vsphere_virtual_machine (ID = <new resource>): Performing diff customization and validation
2018-10-23T11:11:07.071+0100 [DEBUG] plugin.terraform-provider-vsphere_v1.8.1_x4: 2018/10/23 11:11:07 [DEBUG] CdromDiffOperation: Beginning diff validation
2018-10-23T11:11:07.071+0100 [DEBUG] plugin.terraform-provider-vsphere_v1.8.1_x4: 2018/10/23 11:11:07 [DEBUG] cdrom.0 (key 0 at <new device>): Beginning CDROM configuration validation
2018/10/23 11:11:07 [ERROR] root: eval: *terraform.EvalDiff, err: cdrom.0: Either client_device or datastore_id and path must be set
2018/10/23 11:11:07 [ERROR] root: eval: *terraform.EvalSequence, err: cdrom.0: Either client_device or datastore_id and path must be set
2018/10/23 11:11:07 [DEBUG] plugin: waiting for all plugin processes to complete...

Error: Error running plan: 1 error(s) occurred:

* vsphere_virtual_machine.icpmaster: 1 error(s) occurred:

* vsphere_virtual_machine.icpmaster: cdrom.0: Either client_device or datastore_id and path must be set
```
Please provide a link to a GitHub Gist containing the complete debug output:
https://www.terraform.io/docs/internals/debugging.html. Please do NOT paste the
debug output in the issue; just paste a link to the Gist.

### Panic Output
n/a

### Expected Behavior
vsphere_virtual_machine should have accepted `datastore_id` and `path` input
### Actual Behavior
vsphere_virtual_machine complains that `datastore_id` and `path` not provided

### Steps to Reproduce

Please list the steps required to reproduce the issue, for example:
1. Create a template with vsphere_virtual_machine resource including a cdrom section with `datastore_id` and `path` provided.
1. `terraform apply`

### Important Factoids

None that I know

**Comment by bill-rich** [comment]
Hi @hassenius! Thats odd that you're getting that error. It should only be coming up if `datastore_id` or `path` are not set. To help narrow down what is going on here, can you please try filling in `datastore_id` with a hard coded value and also including an output so make sure everything is being found as expected? Something like this should do:

```
output "ds_id" {
  value = "${data.vsphere_datastore.datastore.id}"
}
```

If neither of those turns up anything useful, it would also be helpful to get a full debug log to help track down what is going on.

**Comment by hassenius** [comment]
Thanks @bill-rich 

Did some more digging and testing, and it seems we hit this error when there's some substitution in the path field.
So, this will fail in plan:
```
  cdrom {
    datastore_id = "${data.vsphere_datastore.datastore.id}"
    path         = "cloud-init-iso/tests/${random_id.clusterid.dec}.iso"
  }
```

But this will succeed
```
  cdrom {
    datastore_id = "${data.vsphere_datastore.datastore.id}"
    path         = "cloud-init-iso/tests/1456925035.iso"
  }
```

And this will also succeed
```
  cdrom {
    datastore_id = "${data.vsphere_datastore.datastore.id}"
    path         = "cloud-init-iso/tests/${var.foobar}.iso"
  }
```

Finally, if I apply all the other resources first, then the vsphere_virtual_machine resource will succeed both plan and apply, with the desired substitution in the cdrom block:

```
  cdrom {
    datastore_id = "${data.vsphere_datastore.datastore.id}"
    path         = "cloud-init-iso/tests/${random_id.clusterid.dec}.iso"
  }
```

I think this should make it easily reproducible. Let me know if you want further logs or outputs.

**Comment by bill-rich** [comment]
That is really helpful @hassenius. I believe the issue has to do with validation checks that are run in CustomizeDiff. I'll take a closer look and see if there is a good way to avoid the issue while still performing the cdrom config validation.

**Comment by bill-rich** [comment]
This should be resolved in #784 and released in v1.12.0 of the provider.
